### ISSUE TITLE
32bit arch: EAF_Open error passing argument of 'MA_alloc_get': incompatible pointer type expected Integer {int} but argument is long int

### ISSUE THREAD

**Issue by drew-parsons** [GlobalArrays/ga#352]
The debian ga build has starting failing on 32-bit arches (armel, armhf, i386 etc).  The change in behaviour might be due to gcc-14 tightening compilation conditions, applying `-Wincompatible-pointer-types` to treat pointer inconsistencies as an error.

Build logscan be found at https://buildd.debian.org/status/package.php?p=ga
e.g. [armel](https://buildd.debian.org/status/fetch.php?pkg=ga&arch=armel&ver=5.8.2-4&stamp=1726920401&raw=0)
[i386](https://buildd.debian.org/status/fetch.php?pkg=ga&arch=i386&ver=5.8.2-4&stamp=1726918695&raw=0)  

The error (from armel) is
```
/bin/bash ./libtool  --tag=CC   --mode=compile mpicc.mpich -DHAVE_CONFIG_H -I.         -Ima -I./ma -I./LinAlg/lapack+blas -Iglobal/src -I./global/src -I./global/testing -I./pario/dra -I./pario/eaf -I./pario/elio -I./pario/sf -I./ga++/src  -Igaf2c -I./gaf2c -I./tcgmsg -I./tcgmsg/tcgmsg-mpi       -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2   -fno-aggressive-loop-optimizations -fno-tree-slp-vectorize -g -O2 -c -o pario/eaf/eaf.lo pario/eaf/eaf.c
libtool: compile:  mpicc.mpich -DHAVE_CONFIG_H -I. -Ima -I./ma -I./LinAlg/lapack+blas -Iglobal/src -I./global/src -I./global/testing -I./pario/dra -I./pario/eaf -I./pario/elio -I./pario/sf -I./ga++/src -Igaf2c -I./gaf2c -I./tcgmsg -I./tcgmsg/tcgmsg-mpi -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2 -fno-aggressive-loop-optimizations -fno-tree-slp-vectorize -g -O2 -c pario/eaf/eaf.c -o pario/eaf/eaf.o
pario/eaf/eaf.c: In function 'EAF_Open':
pario/eaf/eaf.c:172:57: error: passing argument 4 of 'MA_alloc_get' from incompatible pointer type [-Wincompatible-pointer-types]
  172 |         if (!MA_alloc_get(MT_CHAR, file[i].size, fname, &handle, &index))
      |                                                         ^~~~~~~
      |                                                         |
      |                                                         long int *
In file included from pario/eaf/eaf.c:63:
ma/macdecls.h:42:18: note: expected 'Integer *' {aka 'int *'} but argument is of type 'long int *'
   42 |     Integer     *memhandle,     /**< RETURN: handle for this block */
      |     ~~~~~~~~~~~~~^~~~~~~~~
pario/eaf/eaf.c:172:66: error: passing argument 5 of 'MA_alloc_get' from incompatible pointer type [-Wincompatible-pointer-types]
  172 |         if (!MA_alloc_get(MT_CHAR, file[i].size, fname, &handle, &index))
      |                                                                  ^~~~~~
      |                                                                  |
      |                                                                  long int *
ma/macdecls.h:43:21: note: expected 'MA_AccessIndex *' {aka 'int *'} but argument is of type 'long int *'
   43 |     MA_AccessIndex *index       /**< RETURN: index for this block */   );
      |     ~~~~~~~~~~~~~~~~^~~~~
make[4]: *** [Makefile:7126: pario/eaf/eaf.lo] Error 1
```

pario/eaf/eaf.c l.172  is using MA_alloc_get  and providing arguments handle and index defined l.145 as long (hardcoded type).
But MA_alloc_get l.42,43 defines the arguments as type Integer.

Integer is defined via gaf2c/typesf2c.h.in l.10 as `@F2C_INTEGER_C_TYPE@`. This is set in configure.ac as long by default, except it gets replaced by m4/ga_f2c_match_types.m4 to the fortran integer matching C int. From the armel build log
```
checking for C type corresponding to INTEGER... int
```

In short, eaf.c hard codes the type as `long`, but MA_alloc_get uses Integer, which is defined as int distinct from long on 32-bit architectures.

If I understand configure correctly (`configure --help`),  I could get a successful 32-bit build by configuring with `--enable-i8`, or alternatively by setting `FFLAG_INT=8` so that Integer matches the `long` expected in eaf.c.  But I'm [ed:] not sure that that's the best resolution of the problem.  Should the hardcoded `long` instead be replaced with `Integer`?

**Comment by jeffhammond** [comment]
> Should the hardcoded long instead be replaced with Integer?

That sounds like the right solution, but I haven't tested it.

**Comment by bjpalmer** [comment]
There is an array of structs defined at the top of the the eaf.c file (called 'file') that has a few elements of type long. I changed these to type Integer and also changed the definition of handle Integer in EAF_Open and changed index to type MA_AccessIndex (the MA_AccessIndex type tracks the size of a void* pointer, independent of the --enable-i8/i4 configuration option). This change compiles without complaint and the eaf test.x problem runs without issues (that I can see).

@edoapra, @jeffhammond are there any tests in NWChem that use eaf? I've checked the fixes into the hotfix/5.9.1 branch.

**Comment by jeffhammond** [comment]
TCE "io eaf" is uncommon but uses EAF.

**Comment by edoapra** [comment]
The DFT code uses eaf. I could test this change in a 32-bit architecture

**Comment by edoapra** [comment]
The new eaf code  works with NWChem on 32-bit ARM

**Comment by drew-parsons** [comment]
The new sparse_array in v5.9 has the same kind of problem,
e.g. https://buildd.debian.org/status/fetch.php?pkg=ga&arch=armhf&ver=5.9-1&stamp=1739999509&raw=0
```
/bin/bash ./libtool  --tag=CC   --mode=compile mpicc.mpich -DHAVE_CONFIG_H -I. -I..         -Ima -I../ma -I../LinAlg/lapack+blas -Iglobal/src -I../global/src -I../global/testing -I../pario/dra -I../pario/eaf -I../pario/elio -I../pario/sf -I../ga++/src  -Igaf2c -I../gaf2c -I../tcgmsg -I../tcgmsg/tcgmsg-mpi   -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2   -fno-aggressive-loop-optimizations -fno-tree-slp-vectorize -std=gnu17 -g -O2 -c -o global/src/sparse.array.lo ../global/src/sparse.array.c
...
libtool: compile:  mpicc.mpich -DHAVE_CONFIG_H -I. -I.. -Ima -I../ma -I../LinAlg/lapack+blas -Iglobal/src -I../global/src -I../global/testing -I../pario/dra -I../pario/eaf -I../pario/elio -I../pario/sf -I../ga++/src -Igaf2c -I../gaf2c -I../tcgmsg -I../tcgmsg/tcgmsg-mpi -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2 -fno-aggressive-loop-optimizations -fno-tree-slp-vectorize -std=gnu17 -g -O2 -c ../global/src/sparse.array.c -o global/src/sparse.array.o
...
../global/src/sparse.array.c: In function 'pnga_sprs_array_matmat_multiply':
../global/src/sparse.array.c:2766:41: error: passing argument 3 of 'pnga_set_data' from incompatible pointer type [-Wincompatible-pointer-types]
 2766 |       pnga_set_data(SPA[hdl_c].g_i,ndim,&totalsize,C_LONG);
      |                                         ^~~~~~~~~~
      |                                         |
      |                                         int64_t * {aka long long int *}
In file included from ../global/src/sparse.array.c:63:
../global/src/ga-papi.h:120:63: note: expected 'Integer *' {aka 'int *'} but argument is of type 'int64_t *' {aka 'long long int *'}
  120 | extern void pnga_set_data(Integer g_a, Integer ndim, Integer *dims, Integer type);
      |                                                      ~~~~~~~~~^~~~
../global/src/sparse.array.c:2768:41: error: passing argument 3 of 'pnga_set_data' from incompatible pointer type [-Wincompatible-pointer-types]
 2768 |       pnga_set_data(SPA[hdl_c].g_i,ndim,&totalsize,C_INT);
      |                                         ^~~~~~~~~~
      |                                         |
      |                                         int64_t * {aka long long int *}
../global/src/ga-papi.h:120:63: note: expected 'Integer *' {aka 'int *'} but argument is of type 'int64_t *' {aka 'long long int *'}
  120 | extern void pnga_set_data(Integer g_a, Integer ndim, Integer *dims, Integer type);
      |                                                      ~~~~~~~~~^~~~
../global/src/sparse.array.c:2799:41: error: passing argument 3 of 'pnga_set_data' from incompatible pointer type [-Wincompatible-pointer-types]
 2799 |       pnga_set_data(SPA[hdl_c].g_j,ndim,&totalsize,C_LONG);
      |                                         ^~~~~~~~~~
      |                                         |
      |                                         int64_t * {aka long long int *}
../global/src/ga-papi.h:120:63: note: expected 'Integer *' {aka 'int *'} but argument is of type 'int64_t *' {aka 'long long int *'}
  120 | extern void pnga_set_data(Integer g_a, Integer ndim, Integer *dims, Integer type);
      |                                                      ~~~~~~~~~^~~~
../global/src/sparse.array.c:2801:41: error: passing argument 3 of 'pnga_set_data' from incompatible pointer type [-Wincompatible-pointer-types]
 2801 |       pnga_set_data(SPA[hdl_c].g_j,ndim,&totalsize,C_INT);
      |                                         ^~~~~~~~~~
      |                                         |
      |                                         int64_t * {aka long long int *}
../global/src/ga-papi.h:120:63: note: expected 'Integer *' {aka 'int *'} but argument is of type 'int64_t *' {aka 'long long int *'}
  120 | extern void pnga_set_data(Integer g_a, Integer ndim, Integer *dims, Integer type);
      |                                                      ~~~~~~~~~^~~~
../global/src/sparse.array.c:2803:42: error: passing argument 3 of 'pnga_set_data' from incompatible pointer type [-Wincompatible-pointer-types]
 2803 |     pnga_set_data(SPA[hdl_c].g_data,ndim,&totalsize,SPA[hdl_c].type);
      |                                          ^~~~~~~~~~
      |                                          |
      |                                          int64_t * {aka long long int *}
../global/src/ga-papi.h:120:63: note: expected 'Integer *' {aka 'int *'} but argument is of type 'int64_t *' {aka 'long long int *'}
  120 | extern void pnga_set_data(Integer g_a, Integer ndim, Integer *dims, Integer type);
      |                                                      ~~~~~~~~~^~~~
libtool: compile:  mpicc.mpich -DHAVE_CONFIG_H -I. -I.. -Ima -I../ma -I../LinAlg/lapack+blas -Iglobal/src -I../global/src -I../global/testing -I../pario/dra -I../pario/eaf -I../pario/elio -I../pario/sf -I../ga++/src -Igaf2c -I../gaf2c -I../tcgmsg -I../tcgmsg/tcgmsg-mpi -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_TIME_BITS=64 -Wdate-time -D_FORTIFY_SOURCE=2 -fno-aggressive-loop-optimizations -fno-tree-slp-vectorize -std=gnu17 -g -O2 -c ../global/src/mtwister.c -o global/src/mtwister.o
make[4]: *** [Makefile:7187: global/src/sparse.array.lo] Error 1
```
Will the type of fix applied to eaf.c work also with sparse.array.c?

**Comment by drew-parsons** [comment]
That's the commit to fix eaf, not sparse.array.
But looks like Bruce has fixed it too, f63e27b
Thanks guys
